docs/root/intro/arch_overview/security/ssl.rst

```diff
@@ -114,21 +114,39 @@ Certificate selection
+  as two different server name patterns.
+* Only one certificate of a particular type (RSA or ECDSA) may be specified for each server name pattern.
 * Non-P-256 server ECDSA certificates are rejected.
-* If the client supports P-256 ECDSA, a P-256 ECDSA certificate will be selected if one is present in the
-  :ref:`DownstreamTlsContext <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`
-  and it is in compliance with the OCSP policy.
-* If the client only supports RSA certificates, a RSA certificate will be selected if present in the
-  :ref:`DownstreamTlsContext <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.
-* Otherwise, the first certificate listed is used. This will result in a failed handshake if the
-  client only supports RSA certificates and the server only has ECDSA certificates.
 * Static and SDS certificates may not be mixed in a given :ref:`DownstreamTlsContext
   <envoy_v3_api_msg_extensions.transport_sockets.tls.v3.DownstreamTlsContext>`.
-* The selected certificate must adhere to the OCSP policy. If no
-  such certificate is found, the connection is refused.
+
+Certificate selection rules:
+
+* If the client supports SNI, e.g. SNI is "test.example.com", it looks for a cert that exactly matches to the SNI.
+* If no cert is selected from certs that exactly matches SNI, it matches on wildcard server name. e.g. if SNI is
+  "test.example.com", a certificate with "test.example.com" will be preferred over "\*.example.com". And wildcard
+  matching only works for 1 level of depth, so "\*.com" will not be a match for "test.example.com".
+* If no cert is found that matches wildcard name, it looks for a cert from whole cert list.
+* It requires OCSP and key type checking in above phases, the candidate cert can be selected only when it adhere to the OCSP policy.
+  It tries to find a cert that can both match OCSP policy and key type, if such cert is not present, the first cert that passing OCSP
+  will be selected even if it does not match to key type.
+* It fallbacks to the first cert in the whole list if there is no cert is selected after above steps.
+* Currently only two kinds of key type are supported, RSA or ECDSA. If the client supports P-256 ECDSA, the P-256 ECDSA certificate
+  is preferred over RSA. The certificate that it falls back to might result in a failed handshake. For instance, a client only supports
+  RSA certificates and the certificate only support ECDSA.
+* The final selected certificate must adhere to the OCSP policy. If no such certificate is found, the connection is refused.
+
+.. note::
+   With the support of SNI-based certificate selection, it allows configuring large number of certificates for multiple hostnames.
+   But if there is no matching SNI or there is OCSP policy failure on the matching SNI, it will look for the cert from the whole cert
+   list(O(n) complexity), which could be a problem for a potential DoS attack.
```

Comment:
I discussed this with @ggreenway and we think the following would be a better/safer default for both DoS safety in Envoy as well as more secure behavior for clients.

1) If no SNI is provided by the client, the behavior should be the same as today. (Adhere to OSCP, etc.)
2) If a client provides SNI, there must be either an exact or wildcard match. If there is no such match, the behavior would be as follows:
  a) Consult a new configuration option which specifies whether to abort the handshake or proceed with a legacy style fulls scan match.
  b) The default for the configuration would be false (abort and do not do a full scan)
  c) we should use a WKT for the boolean, and we should provide a runtime guard that can override the default in case people run into trouble with the default option.

Let me know if this makes sense and @ggreenway lmk if I missed anything.

Reply:
@mattklein123 thanks for your comments

> 1. If no SNI is provided by the client, the behavior should be the same as today. (Adhere to OSCP, etc.)
It can not be exact "the same ". Since we allow loading more than 2 certs with new change, If there is no SNI, this becomes O(n) lookup.  @ggreenway and me considered limit the times of lookup, e.g. we only allow it check the first 2 certs, it can avoid dos issue though it looks not elegant.

> 2. If a client provides SNI, there must be either an exact or wildcard match. If there is no such match, the behavior would be as follows:
>    a) Consult a new configuration option which specifies whether to abort the handshake or proceed with a legacy style fulls scan match.
>    b) The default for the configuration would be false (abort and do not do a full scan)
>    c) we should use a WKT for the boolean, and we should provide a runtime guard that can override the default in case people run into trouble with the default option.

a new configuration option makes sense. What does WKT mean in this context?

Reply:
Sorry WKT means "well known type" so instead of bool you should use https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.BoolValue so you can differentiate between unset/false/true.

Reply:
Thanks.

Are you OK with full scan when no SNI is provided by the client? 

Regarding the the new option, I think it should reside in [extensions.transport_sockets.tls.v3.DownstreamTlsContext](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls.proto#extensions-transport-sockets-tls-v3-downstreamtlscontext), do you agree?
```protobuf
message DownstreamTlsContext {
...
// default is false
google.protobuf.BoolValue full_scan_certs_on_sni_mismatch = 8;
}
```

Reply:
The difference is that previously we said you should only have 2 certs, so full scan was O(2). 

The only realistic case I worry about is if a config only has EC certs, and has LOTS of them, and a client sends a ClientHello specifying only RSA is supported, and no SNI value. We could easily work around that by having a bool computed at config-load for `has_ec` and `has_rsa`, and check that first before even looking for a matching cert (regardless of SNI or not). We can quickly bail out when we know we don't have a compatible cert.

Reply:
> The difference is that previously we said you should only have 2 certs, so full scan was O(2).

Reply:
> We said that but did we check that? If we actually did let people have more than 2 previously maybe it doesn't matter?

We did check that. Envoy would exit when it detected more than 2 certs, at most one ECDSA and at most one RSA was allowed since Envoy only supports these two types. In this PR, we allow at most 2 certs(one ECDSA and one RSA) for each server name pattern, so we might have 2 certs for www.example1.com, 2 for www.example2.com, 2 for *.example.com, and more. Though it is always O(n) lookup, but n <= 2 previously.

Reply:
I'd rather use the existing field. The new behavior is close enough to the old behavior that I don't think we need a new field.

Reply:
Alright sgtm. I guess let's just defer to trying to avoid any full scans by default.

Reply:
@ggreenway @mattklein123 please review the code update, it should meet all requirments.